File: omnimer_health_server/src/domain/models/Exercise/Exercise.model.ts
import mongoose, { Schema, Document, Types } from "mongoose";
import {
  DifficultyLevelEnum,
  DifficultyLevelTuple,
  LocationEnum,
  LocationTuple,
} from "../../../common/constants/EnumConstants";

export interface IExercise extends Document {
  _id: Types.ObjectId;
  name: string;
  description?: string | null;
  instructions?: string;

  equipment: Types.ObjectId[]; // ref: Equipment
  bodyPart: Types.ObjectId[]; // ref: BodyPart
  mainMuscles?: Types.ObjectId[]; // ref: Muscle
  secondaryMuscles?: Types.ObjectId[]; // ref: Muscle
  exerciseTypes: Types.ObjectId[]; // ref: ExerciseType
  exerciseCategories: Types.ObjectId[]; // ref: ExerciseCategory
  location: LocationEnum;

  difficulty?: DifficultyLevelEnum;
  imageUrl?: string | null;
  videoUrl?: string | null;

  rating?: { avg: number; count: number };
}

const exerciseSchema = new Schema<IExercise>(
  {
    _id: { type: Schema.Types.ObjectId, auto: true },

    name: { type: String, required: true },
    description: { type: String, default: null },
    instructions: { type: String, default: null },

    equipment: {
      type: [{ type: Schema.Types.ObjectId, ref: "Equipment" }],
      required: true,
    },
    bodyPart: {
      type: [{ type: Schema.Types.ObjectId, ref: "BodyPart" }],
      required: true,
    },
    mainMuscles: [{ type: Schema.Types.ObjectId, ref: "Muscle" }],
    secondaryMuscles: [{ type: Schema.Types.ObjectId, ref: "Muscle" }],
    exerciseTypes: {
      type: [{ type: Schema.Types.ObjectId, ref: "ExerciseType" }],
      required: true,
    },
    exerciseCategories: {
      type: [{ type: Schema.Types.ObjectId, ref: "ExerciseCategory" }],
      required: true,
    },

    location: {
      type: String,
      enum: LocationTuple,
      default: LocationEnum.None,
    },

    difficulty: {
      type: String,
      enum: DifficultyLevelTuple,
      default: DifficultyLevelEnum.Beginner,
    },

    imageUrl: { type: String, default: null },
    videoUrl: { type: String, default: null },

    rating: {
      type: {
        avg: { type: Number, default: 0 },
        count: { type: Number, default: 0 },
      },
      default: { avg: 0, count: 0 },
    },
  },
  { timestamps: true }
);

export const Exercise = mongoose.model<IExercise>("Exercise", exerciseSchema);

File: omnihealthmobileflutter/lib/data/models/exercise/exercise_model.dart
class ExerciseModel {
  final String id;
  final String name;
  final String image;
  final List<String> mainMuscles;
  final String equipment;
  final String type;
  final String category;
  final String location;

  final String description;
  final String instructions;
  final int? met;
  final double? rating;

  ExerciseModel({
    required this.id,
    required this.name,
    required this.image,
    required this.mainMuscles,
    required this.equipment,
    required this.type,
    required this.category,
    required this.location,
    required this.description,
    required this.instructions,
    this.met,
    this.rating,
  });

  factory ExerciseModel.fromJson(Map<String, dynamic> json) {
    return ExerciseModel(
      id: json['_id'] ?? '',
      name: json['name'] ?? '',
      image: (json['image'] ?? json['imageUrl'] ?? '') as String,
      mainMuscles: List<String>.from(json['mainMuscles'] ?? []),
      equipment: json['equipment'] ?? '',
      type: json['type'] ?? '',
      category: json['category'] ?? '',
      location: json['location'] ?? '',
      description: json['description'] ?? '',
      instructions: json['instructions'] ?? '',
      met: json['met'] == null ? null : (json['met'] as num).toInt(),
      rating: json['rating'] == null
          ? null
          : (json['rating'] as num).toDouble(),
    );
  }

  ExerciseModel copyWith({
    String? id,
    String? name,
    String? image,
    List<String>? mainMuscles,
    String? equipment,
    String? type,
    String? category,
    String? location,
    String? description,
    String? instructions,
    int? met,
    double? rating,
  }) {
    return ExerciseModel(
      id: id ?? this.id,
      name: name ?? this.name,
      image: image ?? this.image,
      mainMuscles: mainMuscles ?? this.mainMuscles,
      equipment: equipment ?? this.equipment,
      type: type ?? this.type,
      category: category ?? this.category,
      location: location ?? this.location,
      description: description ?? this.description,
      instructions: instructions ?? this.instructions,
      met: met ?? this.met,
      rating: rating ?? this.rating,
    );
  }
}
